lib/battle/DamageCalculator.cpp

```diff
@@ -0,0 +1,518 @@
+}
+
+double DamageCalculator::getDefenseMagicFactor() const
+{
+	// Magic Elementals deal half damage (R8 = 0.50) against Magic Elementals and Black Dragons. This is not affected by the Orb of Vulnerability, Anti-Magic, or Magic Resistance.
+	if(info.attacker->creatureIndex() == CreatureID::MAGIC_ELEMENTAL)
+	{
+		const std::string cachingStrMagicImmunity = "type_LEVEL_SPELL_IMMUNITY";
+		static const auto selectorMagicImmunity = Selector::type()(Bonus::LEVEL_SPELL_IMMUNITY);
+
+		if(info.defender->valOfBonuses(selectorMagicImmunity, cachingStrMagicImmunity) >= 5)
+			return 0.5;
+	}
+	return 0.0;
+}
+
+double DamageCalculator::getDefenseMindFactor() const
+{
+	// Psychic Elementals deal half damage (R8 = 0.50) against creatures that are immune to Mind spells, such as Giants and Undead. This is not affected by the Orb of Vulnerability.
+	if(info.attacker->creatureIndex() == CreatureID::PSYCHIC_ELEMENTAL)
+	{
+		const std::string cachingStrMindImmunity = "type_MIND_IMMUNITY";
+		static const auto selectorMindImmunity = Selector::type()(Bonus::MIND_IMMUNITY);
+
+		if(info.defender->hasBonus(selectorMindImmunity, cachingStrMindImmunity))
+			return 0.5;
+	}
+	return 0.0;
+}
+
+std::vector<double> DamageCalculator::getAttackFactors() const
+{
+	return {
+		getAttackSkillFactor(),
+		getAttackOffenseArcheryFactor(),
+		getAttackBlessFactor(),
+		getAttackLuckFactor(),
+		getAttackJoustingFactor(),
+		getAttackDeathBlowFactor(),
+		getAttackDoubleDamageFactor(),
+		getAttackHateFactor()
+	};
+}
+
+std::vector<double> DamageCalculator::getDefenseFactors() const
+{
+	return {
+		getDefenseSkillFactor(),
+		getDefenseArmorerFactor(),
+		getDefenseMagicShieldFactor(),
+		getDefenseRangePenaltiesFactor(),
+		getDefenseObstacleFactor(),
+		getDefenseBlindParalysisFactor(),
+		getDefenseUnluckyFactor(),
+		getDefenseForgetfulnessFactor(),
+		getDefensePetrificationFactor(),
+		getDefenseMagicFactor(),
+		getDefenseMindFactor()
+	};
+}
+
+int DamageCalculator::battleBonusValue(const IBonusBearer * bearer, const CSelector & selector) const
+{
+	auto noLimit = Selector::effectRange()(Bonus::NO_LIMIT);
+	auto limitMatches = info.shooting
+						? Selector::effectRange()(Bonus::ONLY_DISTANCE_FIGHT)
+						: Selector::effectRange()(Bonus::ONLY_MELEE_FIGHT);
+
+	//any regular bonuses or just ones for melee/ranged
+	return bearer->getBonuses(selector, noLimit.Or(limitMatches))->totalValue();
+};
+
+TDmgRange DamageCalculator::calculateDmgRange() const
+{
+	TDmgRange result = getBaseDamageStack();
+
+	auto attackFactors = getAttackFactors();
+	auto defenseFactors = getDefenseFactors();
+
+	double attackFactorTotal = 1.0;
+	double defenseFactorTotal = 1.0;
+
+	for (auto & factor : attackFactors)
+	{
+		assert(factor >= 0.0);
+		attackFactorTotal += factor;
+	}
+
+	for (auto & factor : defenseFactors)
+	{
+		assert(factor >= 0.0);
+		defenseFactorTotal *= ( 1 - std::min(1.0, factor));
```

Comment:
@DjWarmonger overflow check is located here

Reply:
Okay then.

I'm pretty sure many other bonuses can behave odd with some overflow, but that's the topic for another day.